File: environment/src/main/java/jetbrains/exodus/tree/btree/LeafNodeDupMutable.java
/**
 * Copyright 2010 - 2018 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.tree.btree;

import jetbrains.exodus.ArrayByteIterable;
import jetbrains.exodus.ByteIterable;
import jetbrains.exodus.tree.ITree;
import org.jetbrains.annotations.NotNull;

import java.io.PrintStream;

/**
 * Stateful leaf node with root page of duplicates sub-tree as a value
 */
class LeafNodeDupMutable extends BaseLeafNodeMutable {

    @NotNull
    protected final BTreeDupMutable tree;

    LeafNodeDupMutable(@NotNull final BTreeDupMutable tree) {
        this.tree = tree;
    }

    @Override
    public long getAddress() {
        return tree.address;
    }

    @NotNull
    @Override
    public BTreeBase getTree() {
        return tree;
    }

    @NotNull
    @Override
    public AddressIterator addressIterator() {
        final BTreeTraverser traverser = BTreeMutatingTraverser.create(tree);
        return new AddressIterator(null, traverser.currentNode.size > 0, traverser);
    }

    @Override
    public boolean hasValue() {
        return false;
    }

    @Override
    public boolean isDup() {
        return true;
    }

    @Override
    public long getDupCount() {
        return tree.size;
    }

    @NotNull
    BasePageMutable getRootPage() {
        return tree.getRoot();
    }

    @Override
    public boolean valueExists(@NotNull ByteIterable value) {
        // value is a key in duplicates sub-tree
        return tree.hasKey(value);
    }

    @Override
    public int compareKeyTo(@NotNull final ByteIterable iterable) {
        return tree.key.compareTo(iterable);
    }

    @Override
    public int compareValueTo(@NotNull final ByteIterable iterable) {
        return getValue().compareTo(iterable);
    }

    @Override
    @NotNull
    public ByteIterable getKey() {
        return tree.key;
    }

    @Override
    @NotNull
    public ByteIterable getValue() {
        return tree.getRoot().getMinKey().getKey();
    }

    @Override
    public boolean delete(ByteIterable value) {
        return tree.delete(value);
    }

    boolean put(@NotNull ByteIterable value) {
        return tree.put(value, ByteIterable.EMPTY);
    }

    LeafNodeDupMutable putRight(@NotNull ByteIterable value) {
        tree.putRight(value, ArrayByteIterable.EMPTY);
        return this;
    }

    @Override
    public long save(final ITree mainTree) {
        if (tree.mainTree != mainTree) {
            throw new IllegalArgumentException("Can't save LeafNodeDupMutable against mutable tree " +
                    "different from passed on creation");
        }
        return tree.save();
    }

    @Override
    public String toString() {
        return "LND* {key:" + getKey().toString() + '}';
    }

    @Override
    public void dump(PrintStream out, int level, ToString renderer) {
        super.dump(out, level, renderer);
        tree.getRoot().dump(out, level + 1, renderer);
    }

    /**
     * Convert any leaf to mutable leaf with duplicates support
     *
     * @param ln       leaf node to convert
     * @param mainTree its tree
     * @return mutable copy of ln
     */
    static LeafNodeDupMutable convert(@NotNull ILeafNode ln, @NotNull BTreeMutable mainTree) {

        final boolean isLeafNodeDup = ln.isDup();
        if (isLeafNodeDup && ln instanceof LeafNodeDupMutable) {
            return (LeafNodeDupMutable) ln;
        }

        // wrapper tree that doesn't allow duplicates
        final BTreeDupMutable dupTree = isLeafNodeDup ?
                ((LeafNodeDup) ln).getTreeCopyMutable() :
                new BTreeDupMutable(
                        new BTreeEmpty(mainTree.getLog(), mainTree.getBalancePolicy(), false, mainTree.getStructureId()),
                        ln.getKey()
                );
        dupTree.mainTree = mainTree;
        return convert(ln, mainTree, dupTree);
    }

    static LeafNodeDupMutable convert(@NotNull ILeafNode ln, @NotNull BTreeMutable mainTree, @NotNull BTreeDupMutable dupTree) {
        final LeafNodeDupMutable result = new LeafNodeDupMutable(dupTree);
        if (ln.isDup()) {
            return result;
        } else {
            // leaf node with one value -- add it
            mainTree.decrementSize(1); // hack
            result.put(ln.getValue());
            return result;
        }
    }
}

File: environment/src/main/kotlin/jetbrains/exodus/tree/btree/LeafNodeDupMutable.kt
/*
 * Copyright 2010 - 2023 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.tree.btree

import jetbrains.exodus.*
import jetbrains.exodus.tree.Dumpable
import jetbrains.exodus.tree.ITree
import java.io.PrintStream

/**
 * Stateful leaf node with root page of duplicates sub-tree as a value
 */
internal class LeafNodeDupMutable(@JvmField val tree: BTreeDupMutable) : BaseLeafNodeMutable() {
    override fun getAddress(): Long = tree.address

    override fun getTree(): BTreeBase = tree

    override fun addressIterator(): AddressIterator {
        val traverser: BTreeTraverser = BTreeMutatingTraverser.create(tree)
        return AddressIterator(null, traverser.currentNode.size > 0, traverser)
    }

    override fun hasValue(): Boolean {
        return false
    }

    override fun isDup(): Boolean = true
    override fun getDupCount(): Long = tree.size
    fun getRootPage(): BasePageMutable = tree.getRoot()

    override fun valueExists(value: ByteIterable): Boolean {
        // value is a key in duplicates sub-tree
        return tree.hasKey(value)
    }

    override fun compareKeyTo(iterable: ByteIterable): Int {
        return tree.key.compareTo(iterable)
    }

    override fun compareValueTo(iterable: ByteIterable): Int {
        return getValue().compareTo(iterable)
    }

    override fun getKey(): ByteIterable = tree.key
    override fun getValue(): ByteIterable = tree.root.getMinKey().getKey()

    override fun delete(value: ByteIterable?): Boolean {
        return tree.delete(value!!)
    }

    fun put(value: ByteIterable): Boolean {
        return tree.put(value, ByteIterable.EMPTY)
    }

    fun putRight(value: ByteIterable): LeafNodeDupMutable {
        tree.putRight(value, ArrayByteIterable.EMPTY)
        return this
    }

    override fun save(tree: ITree): Long {
        require(this.tree.mainTree === tree) {
            "Can't save LeafNodeDupMutable against mutable tree " +
                    "different from passed on creation"
        }
        return this.tree.save()
    }

    override fun toString(): String {
        return "LND* {key:${getKey()}}"
    }

    override fun dump(out: PrintStream, level: Int, renderer: Dumpable.ToString?) {
        super.dump(out, level, renderer)
        tree.root.dump(out, level + 1, renderer)
    }

    companion object {
        /**
         * Convert any leaf to mutable leaf with duplicates support
         *
         * @param ln       leaf node to convert
         * @param mainTree its tree
         * @return mutable copy of ln
         */
        fun convert(ln: ILeafNode, mainTree: BTreeMutable): LeafNodeDupMutable {
            val isLeafNodeDup = ln.isDup()
            if (isLeafNodeDup && ln is LeafNodeDupMutable) {
                return ln
            }

            // wrapper tree that doesn't allow duplicates
            val dupTree = if (isLeafNodeDup) (ln as LeafNodeDup).getTreeCopyMutable() else BTreeDupMutable(
                BTreeEmpty(mainTree.log, mainTree.balancePolicy, false, mainTree.structureId),
                ln.getKey()
            )
            dupTree.mainTree = mainTree
            return convert(ln, mainTree, dupTree)
        }

        fun convert(ln: ILeafNode, mainTree: BTreeMutable, dupTree: BTreeDupMutable): LeafNodeDupMutable {
            val result = LeafNodeDupMutable(dupTree)
            return if (ln.isDup()) {
                result
            } else {
                // leaf node with one value -- add it
                mainTree.decrementSize(1) // hack
                result.put(ln.getValue()!!)
                result
            }
        }
    }
}
